pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/relocation/RealtimeSegmentRelocator.java

```diff
@@ -0,0 +1,278 @@
+    List<String> allRealtimeTableNames = _pinotHelixResourceManager.getAllRealtimeTables();
+
+    for (final String tableNameWithType : allRealtimeTableNames) {
+      try {
+        LOGGER.info("Starting relocation of segments for table: {}", tableNameWithType);
+
+        TableConfig tableConfig = _pinotHelixResourceManager.getRealtimeTableConfig(tableNameWithType);
+        final RealtimeTagConfig realtimeTagConfig = new RealtimeTagConfig(tableConfig, _helixManager);
+        if (!realtimeTagConfig.isRelocateCompletedSegments()) {
+          LOGGER.info("Skipping relocation of segments for {}", tableNameWithType);
+          return;
+        }
+
+        Function<IdealState, IdealState> updater = new Function<IdealState, IdealState>() {
+          @Nullable
+          @Override
+          public IdealState apply(@Nullable IdealState idealState) {
+            if (!idealState.isEnabled()) {
+              LOGGER.info("Skipping relocation of segments for {} since ideal state is disabled", tableNameWithType);
+              return null;
+            }
+            relocateSegments(realtimeTagConfig, idealState);
+            return idealState;
+          }
+        };
+
+        HelixHelper.updateIdealState(_helixManager, tableNameWithType, updater,
+            RetryPolicies.exponentialBackoffRetryPolicy(5, 1000, 2.0f));
+      } catch (Exception e) {
+        LOGGER.error("Exception in relocating realtime segments of table {}", tableNameWithType, e);
+      }
+    }
+    LOGGER.info("Realtime segment relocation completed");
+  }
+
+  /**
+   * Given a realtime tag config and an ideal state, relocate the segments
+   * which are completed but still hanging around on consuming servers, one replica at a time
+   * @param  realtimeTagConfig
+   * @param idealState
+   */
+  protected void relocateSegments(RealtimeTagConfig realtimeTagConfig, IdealState idealState) {
+
+    List<String> consumingServers = _helixAdmin.getInstancesInClusterWithTag(_helixManager.getClusterName(),
+        realtimeTagConfig.getConsumingServerTag());
+    if (consumingServers.isEmpty()) {
+      throw new IllegalStateException(
+          "Found no realtime consuming servers with tag " + realtimeTagConfig.getConsumingServerTag());
+    }
+    List<String> completedServers = _helixAdmin.getInstancesInClusterWithTag(_helixManager.getClusterName(),
+        realtimeTagConfig.getCompletedServerTag());
+    if (completedServers.isEmpty()) {
+      throw new IllegalStateException(
+          "Found no realtime completed servers with tag " + realtimeTagConfig.getCompletedServerTag());
+    }
+
+    // TODO: use segment assignment strategy to decide where to place relocated segment
+
+    // create map of completed server name to num segments it holds.
+    // This will help us decide which completed server to choose for replacing a consuming server
+    Map<String, Integer> completedServerToNumSegments = new HashMap<>(completedServers.size());
```

Comment:
will be good to have extract this logic into an interface, we might come up with different implementations and will abstract out the logic. The interface should basically take an idealstate, set of consuming nodes, set of completed nodes and return a new idealstate.

Reply:
I don't plan to keep this logic for long (see TODO). The plan is to be use SegmentAssignmentStrategy for this assignment. Currently the interface and implementations of SegmentAssignmentStrategy aren't directly usable for for this purpose. Will be working on that in another PR.